app/src/main/java/com/example/drawingactivity/GradientView.kt

```diff
@@ -0,0 +1,166 @@
+import android.content.Context
+import android.graphics.*
+import android.graphics.Color.BLACK
+import android.graphics.Color.HSVToColor
+import android.util.AttributeSet
+import android.view.MotionEvent
+import android.view.View
+import com.example.drawingactivity.ColorConstants.hueMaxFloat
+import kotlinx.coroutines.*
+
+@SuppressLint("ClickableViewAccessibility")
+class GradientView @JvmOverloads constructor(
+    context: Context,
+    attrs: AttributeSet? = null,
+    defStyle: Int = 0
+) : View(context, attrs, defStyle) {
+
+    var onSatValChanged: ((Float, Float) -> Unit)? = null
+    var onGradientClicked: ((Boolean) -> Unit)? = null
+    private var currentHue = hueMaxFloat
+    private var shape: Rect? = null
+    private var dx = 0f
+    private var dy = 0f
+
+    init {
+        setOnTouchListener { _, mEvent ->
+            when (mEvent.action) {
+                MotionEvent.ACTION_DOWN -> {
+                    onGradientClicked?.invoke(true)
+                    dx = mEvent.x.coerceIn(0f, width * 1f)
+                    dy = mEvent.y.coerceIn(0f, height * 1f)
+                    if (dy > height / 2) {
+                        selectPaint.color = Color.WHITE
+                    } else selectPaint.color = BLACK
+                    onSatValChanged?.invoke(dx / width, 1 - (dy / height))
+                    invalidate()
+                }
+                MotionEvent.ACTION_MOVE, MotionEvent.ACTION_CANCEL -> {
+                    dx = mEvent.x.coerceIn(0f, width * 1f)
+                    dy = mEvent.y.coerceIn(0f, height * 1f)
+                    if (dy > height / 2) {
+                        selectPaint.color = Color.WHITE
+                    } else selectPaint.color = BLACK
+                    onSatValChanged?.invoke(dx / width, 1 - (dy / height))
+                    invalidate()
+                }
+                MotionEvent.ACTION_UP -> {
+                    onGradientClicked?.invoke(false)
+                }
+            }
+            true
+        }
+    }
+
+    fun setHue(huePickerValue: Float) {
+        currentHue = huePickerValue
+        invalidate()
+    }
+
+    fun setSaturationAndValue(saturation: Float, value: Float) {
+        dx = saturation * width
+        dy = value * height
+        invalidate()
+    }
+
+    override fun onDraw(canvas: Canvas?) {
+        super.onDraw(canvas)
+        canvas?.apply {
+            if (shape == null) {
+                shape = Rect(0, 0, width, height)
+            }
+            generateAllBitmaps()
+            getBitmap(currentHue.toInt())?.let { bitmap ->
+                shape?.let { shape ->
+                    drawBitmap(
+                        bitmap,
+                        null,
+                        shape,
+                        null
+                    )
+                }
+            }
+            drawCircle(dx, height - dy, 12f, selectPaint)
+        }
+    }
+
+    companion object GradientConstants {
```

Comment:
Don't ever need to name your companions. If you feel the need to name them they should be separate

Reply:
they are now namelets